--- a/cpp/Makefile
+++ b/cpp/Makefile
@@ -11,11 +11,7 @@
 
 include $(top_srcdir)/config/Make.rules
 
-SUBDIRS		= config src include test
-
-ifeq ($(shell uname | grep MINGW),)
-SUBDIRS		:= $(SUBDIRS) demo
-endif
+SUBDIRS		= config src include
 
 INSTALL_SUBDIRS	= $(install_bindir) $(install_libdir)$(cpp11libdirsuffix) $(install_includedir) \
 	$(install_configdir) $(install_mandir)
--- a/cs/Makefile
+++ b/cs/Makefile
@@ -11,7 +11,7 @@
 
 include $(top_srcdir)/config/Make.rules.cs
 
-SUBDIRS		= src test demo
+SUBDIRS		= src
 
 INSTALL_SUBDIRS	= $(install_bindir) $(install_assembliesdir) $(install_mandir)
 
--- a/js/Makefile
+++ b/js/Makefile
@@ -11,7 +11,7 @@
 
 include $(top_srcdir)/config/Make.rules.js
 
-SUBDIRS		= src assets test demo
+SUBDIRS		= src assets
 
 INSTALL_SUBDIRS	= $(install_bindir) $(install_libdir) $(install_moduledir)
 
--- a/php/Makefile
+++ b/php/Makefile
@@ -11,7 +11,7 @@
 
 include $(top_srcdir)/config/Make.rules.php
 
-SUBDIRS		= src lib demo test
+SUBDIRS		= src lib
 INSTALL_SUBDIRS = $(install_phpdir) $(install_libdir)
 
 install:: install-common
--- a/py/Makefile
+++ b/py/Makefile
@@ -11,7 +11,7 @@
 
 include $(top_srcdir)/config/Make.rules
 
-SUBDIRS		= modules python test
+SUBDIRS		= modules python
 
 INSTALL_SUBDIRS = $(install_pythondir) $(install_libdir)
 
--- a/Makefile
+++ b/Makefile
@@ -7,10 +7,10 @@
 #
 # **********************************************************************
 
-SUBDIRS			= cpp java cs js py rb php
-CLEAN_SUBDIRS		= js java cs py rb php cpp
-DEPEND_SUBDIRS		= cpp cs py rb php
-INSTALL_SUBDIRS		= cpp java cs py rb php js
+SUBDIRS			= cpp java js py rb php
+CLEAN_SUBDIRS		= js java py rb php cpp
+DEPEND_SUBDIRS		= cpp py rb php
+INSTALL_SUBDIRS		= cpp java py rb php js
 
 all::
 	@for subdir in $(SUBDIRS); \
--- a/config/Make.common.rules
+++ b/config/Make.common.rules
@@ -79,7 +79,10 @@
         cpp11libdirsuffix	= /c++11
         cpp11libsuffix		= ++11
         cpp11sonamedir		= ../
+        cppsubdir		= cpp11
     endif
+else
+    cppsubdir			= cpp
 endif
 
 ifneq ($(findstring MINGW,$(UNAME)),)
@@ -134,16 +137,16 @@
 		ifneq ($(shell test -f $(ICE_HOME)/$(binsubdir)/$(slice_translator) && echo 0), 0)
 $(error Unable to find $(slice_translator) in $(ICE_HOME)/$(binsubdir), please verify ICE_HOME is properly configured and Ice is correctly installed.)
 		endif
-		ifeq ($(shell test -f $(ice_dir)/cpp/bin/$(slice_translator) && echo 0), 0)
-$(warning Found $(slice_translator) in both ICE_HOME/bin and $(ice_dir)/cpp/bin, ICE_HOME/bin/$(slice_translator) will be used!)
+		ifeq ($(shell test -f $(ice_dir)/$(cppsubdir)/bin/$(slice_translator) && echo 0), 0)
+$(warning Found $(slice_translator) in both ICE_HOME/bin and $(ice_dir)/$(cppsubdir)/bin, ICE_HOME/bin/$(slice_translator) will be used!)
 		endif
 		ice_cpp_dir = $(ICE_HOME)
 	    else
 $(warning Ignoring ICE_HOME environment variable to build current source tree.)
-		ice_cpp_dir = $(ice_dir)/cpp
+		ice_cpp_dir = $(ice_dir)/$(cppsubdir)
 	    endif
 	else
-	    ice_cpp_dir = $(ice_dir)/cpp
+	    ice_cpp_dir = $(ice_dir)/$(cppsubdir)
 	endif
 
     endif
@@ -162,16 +165,16 @@
 		ifneq ($(shell test -f $(ICE_HOME)/$(binsubdir)/$(slice_translator) && echo 0), 0)
 $(error Unable to find $(slice_translator) in $(ICE_HOME)/$(binsubdir), please verify ICE_HOME is properly configured and Ice is correctly installed.)
 		endif
-		ifeq ($(shell test -f $(ice_dir)/cpp/bin/$(slice_translator) && echo 0), 0)
-$(warning Found $(slice_translator) in both ICE_HOME/bin and $(ice_dir)/cpp/bin, ICE_HOME/bin/$(slice_translator) will be used!)
+		ifeq ($(shell test -f $(ice_dir)/$(cppsubdir)/bin/$(slice_translator) && echo 0), 0)
+$(warning Found $(slice_translator) in both ICE_HOME/bin and $(ice_dir)/$(cppsubdir)/bin, ICE_HOME/bin/$(slice_translator) will be used!)
 		endif
 		ice_cpp_dir = $(ICE_HOME)
 	    else
 $(warning Ignoring ICE_HOME environment variable to build current source tree.)
-		ice_cpp_dir = $(ice_dir)/cpp
+		ice_cpp_dir = $(ice_dir)/$(cppsubdir)
 	    endif
 	else
-	    ice_cpp_dir = $(ice_dir)/cpp
+	    ice_cpp_dir = $(ice_dir)/$(cppsubdir)
 	endif
     endif
 endif
@@ -262,7 +265,7 @@
 #
 ifneq ($(ice_dir), /usr)
     ifdef ice_src_dist
-	ifeq ($(ice_cpp_dir), $(ice_dir)/cpp)
+	ifeq ($(ice_cpp_dir), $(ice_dir)/$(cppsubdir))
 	    ice_lib_dir = $(ice_cpp_dir)/lib
 	else
 	ice_lib_dir = $(ice_cpp_dir)/$(libsubdir)
--- a/java/Makefile
+++ b/java/Makefile
@@ -16,17 +16,17 @@
 endif
 
 all:
-	./gradlew assemble
+	./gradlew -g$(HOME) assemble
 
 dist:
-	./gradlew :Ice:assemble :Freeze:assemble :Glacier2:assemble :IceGrid:assemble :ant:assemble \
+	./gradlew -g$(HOME) :Ice:assemble :Freeze:assemble :Glacier2:assemble :IceGrid:assemble :ant:assemble \
 		:IceBox:assemble :IceDiscovery:assemble :IcePatch2:assemble :IceStorm:assemble :IceGridGUI:assemble
 
 clean:
-	./gradlew clean
+	./gradlew -g$(HOME) clean
 
 install::
-	./gradlew $(INSTALL_FLAGS) install
+	./gradlew -g$(HOME) $(INSTALL_FLAGS) install
 
 test::
 	@python ./allTests.py
--- a/cpp/config/Make.rules
+++ b/cpp/config/Make.rules
@@ -17,7 +17,7 @@
 # The "root directory" for runpath embedded in executables. Can be unset
 # to avoid adding a runpath to Ice executables.
 #
-embedded_runpath_prefix ?= /opt/Ice-$(VERSION_MAJOR).$(VERSION_MINOR)
+#embedded_runpath_prefix ?= /opt/Ice-$(VERSION_MAJOR).$(VERSION_MINOR)
 
 #
 # Define create_runpath_symlink as yes if you would like 'make install'
