--- a/config/Make.common.rules
+++ b/config/Make.common.rules
@@ -95,7 +95,10 @@
         cpp11libsuffix		= ++11
         cpp11sonamedir		= ../
         binsuffix 		:= $(binsuffix)$(cpp11libsuffix)
+        cppsubdir 		= cpp11
     endif
+else
+    cppsubdir			= cpp
 endif
 
 ifneq ($(findstring MINGW,$(UNAME)),)
@@ -165,16 +168,16 @@
 		ifneq ($(shell test -f $(ICE_HOME)/$(binsubdir)/$(slice_translator) && echo 0), 0)
 $(error Unable to find $(slice_translator) in $(ICE_HOME)/$(binsubdir), please verify ICE_HOME is properly configured and Ice is correctly installed.)
 		endif
-		ifeq ($(shell test -f $(ice_dir)/cpp/bin/$(slice_translator) && echo 0), 0)
-$(warning Found $(slice_translator) in both ICE_HOME/bin and $(ice_dir)/cpp/bin, ICE_HOME/bin/$(slice_translator) will be used!)
+		ifeq ($(shell test -f $(ice_dir)/$(cppsubdir)/bin/$(slice_translator) && echo 0), 0)
+$(warning Found $(slice_translator) in both ICE_HOME/bin and $(ice_dir)/$(cppsubdir)/bin, ICE_HOME/bin/$(slice_translator) will be used!)
 		endif
 		ice_cpp_dir = $(ICE_HOME)
 	    else
 $(warning Ignoring ICE_HOME environment variable to build current source tree.)
-		ice_cpp_dir = $(ice_dir)/cpp
+		ice_cpp_dir = $(ice_dir)/$(cppsubdir)
 	    endif
 	else
-	    ice_cpp_dir = $(ice_dir)/cpp
+	    ice_cpp_dir = $(ice_dir)/$(cppsubdir)
 	endif
 
     endif
@@ -193,16 +196,16 @@
 		ifneq ($(shell test -f $(ICE_HOME)/$(binsubdir)/$(slice_translator) && echo 0), 0)
 $(error Unable to find $(slice_translator) in $(ICE_HOME)/$(binsubdir), please verify ICE_HOME is properly configured and Ice is correctly installed.)
 		endif
-		ifeq ($(shell test -f $(ice_dir)/cpp/bin/$(slice_translator) && echo 0), 0)
-$(warning Found $(slice_translator) in both ICE_HOME/bin and $(ice_dir)/cpp/bin, ICE_HOME/bin/$(slice_translator) will be used!)
+		ifeq ($(shell test -f $(ice_dir)/$(cppsubdir)/bin/$(slice_translator) && echo 0), 0)
+$(warning Found $(slice_translator) in both ICE_HOME/bin and $(ice_dir)/$(cppsubdir)/bin, ICE_HOME/bin/$(slice_translator) will be used!)
 		endif
 		ice_cpp_dir = $(ICE_HOME)
 	    else
 $(warning Ignoring ICE_HOME environment variable to build current source tree.)
-		ice_cpp_dir = $(ice_dir)/cpp
+		ice_cpp_dir = $(ice_dir)/$(cppsubdir)
 	    endif
 	else
-	    ice_cpp_dir = $(ice_dir)/cpp
+	    ice_cpp_dir = $(ice_dir)/$(cppsubdir)
 	endif
     endif
 endif
--- a/cpp/Makefile
+++ b/cpp/Makefile
@@ -13,13 +13,6 @@
 
 SUBDIRS		= config src include
 
-ifneq ($(MAKECMDGOALS),install)
-    SUBDIRS		:= $(SUBDIRS) test
-    ifeq ($(shell uname | grep MINGW),)
-        SUBDIRS		:= $(SUBDIRS) demo
-    endif
-endif
-
 INSTALL_SUBDIRS	= $(install_bindir) $(install_libdir)$(cpp11libdirsuffix) $(install_includedir) \
 	$(install_configdir) $(install_mandir)
 
--- a/php/Makefile
+++ b/php/Makefile
@@ -13,10 +13,6 @@
 
 SUBDIRS		= src lib
 
-ifneq ($(MAKECMDGOALS),install)
-SUBDIRS		:= $(SUBDIRS) test demo
-endif
-
 INSTALL_SUBDIRS = $(install_phpdir) $(install_libdir)
 
 install:: install-common
--- a/py/Makefile
+++ b/py/Makefile
@@ -13,10 +13,6 @@
 
 SUBDIRS		= modules python
 
-ifneq ($(MAKECMDGOALS),install)
-SUBDIRS		:= $(SUBDIRS) test
-endif
-
 INSTALL_SUBDIRS = $(install_pythondir) $(install_libdir)
 
 install:: install-common
--- a/java/Makefile
+++ b/java/Makefile
@@ -20,9 +20,6 @@
 endif
 
 all:
-	$(GRADLE) build
-
-dist:
 	$(GRADLE) :Ice:assemble :Freeze:assemble :Glacier2:assemble :IceGrid:assemble :ant:assemble \
 		:IceBox:assemble :IceDiscovery:assemble :IcePatch2:assemble :IceStorm:assemble :IceGridGUI:assemble
 
--- a/Makefile
+++ b/Makefile
@@ -7,10 +7,10 @@
 #
 # **********************************************************************
 
-SUBDIRS			= cpp java js py rb php
-CLEAN_SUBDIRS	= js java py rb php cpp
+SUBDIRS			= cpp java php
+CLEAN_SUBDIRS	= java php cpp
 DEPEND_SUBDIRS	= cpp py rb php
-INSTALL_SUBDIRS	= cpp java py rb php js
+INSTALL_SUBDIRS	= cpp java php
 
 ifeq ($(shell uname),Darwin)
 SUBDIRS			+= objc
