--- a/config/Make.common.rules
+++ b/config/Make.common.rules
@@ -121,7 +121,10 @@
         cpp11libsuffix		= ++11
         cpp11sonamedir		= ../
         binsuffix 		:= $(binsuffix)$(cpp11libsuffix)
+        cppsubdir 		= cpp11
     endif
+else
+    cppsubdir			= cpp
 endif
 
 ifneq ($(findstring MINGW,$(UNAME)),)
@@ -189,7 +192,7 @@
 	        #
 		# ICE_HOME is pointing at the source distribution, so we just ignore it.
 		#
-	        ice_cpp_dir = $(ice_dir)/cpp
+	        ice_cpp_dir = $(ice_dir)/$(cppsubdir)
 	    else
 	        ifdef slice_translator
 		    ifneq ($(shell test -f $(ICE_HOME)/$(binsubdir)/$(slice_translator) && echo 0), 0)
@@ -201,11 +204,11 @@
 		    ice_cpp_dir = $(ICE_HOME)
 	        else
 $(warning Ignoring ICE_HOME environment variable to build current source tree.)
-		    ice_cpp_dir = $(ice_dir)/cpp
+		    ice_cpp_dir = $(ice_dir)/$(cppsubdir)
 	        endif
 	    endif
 	else
-	    ice_cpp_dir = $(ice_dir)/cpp
+	    ice_cpp_dir = $(ice_dir)/$(cppsubdir)
 	endif
 
     endif
@@ -224,16 +227,16 @@
 		ifneq ($(shell test -f $(ICE_HOME)/$(binsubdir)/$(slice_translator) && echo 0), 0)
 $(error Unable to find $(slice_translator) in $(ICE_HOME)/$(binsubdir), please verify ICE_HOME is properly configured and Ice is correctly installed.)
 		endif
-		ifeq ($(shell test -f $(ice_dir)/cpp/bin/$(slice_translator) && echo 0), 0)
-$(warning Found $(slice_translator) in both ICE_HOME/bin and $(ice_dir)/cpp/bin, ICE_HOME/bin/$(slice_translator) will be used!)
+		ifeq ($(shell test -f $(ice_dir)/$(cppsubdir)/bin/$(slice_translator) && echo 0), 0)
+$(warning Found $(slice_translator) in both ICE_HOME/bin and $(ice_dir)/$(cppsubdir)/bin, ICE_HOME/bin/$(slice_translator) will be used!)
 		endif
 		ice_cpp_dir = $(ICE_HOME)
 	    else
 $(warning Ignoring ICE_HOME environment variable to build current source tree.)
-		ice_cpp_dir = $(ice_dir)/cpp
+		ice_cpp_dir = $(ice_dir)/$(cppsubdir)
 	    endif
 	else
-	    ice_cpp_dir = $(ice_dir)/cpp
+	    ice_cpp_dir = $(ice_dir)/$(cppsubdir)
 	endif
     endif
 endif
@@ -327,7 +330,7 @@
 #
 # Set environment variables for the Slice translator.
 #
-ifneq ($(ice_dir), $(usr_dir))
+ifneq ($(ice_dir),$(usr_dir))
     ifdef ice_src_dist
         ice_lib_dir = $(ice_cpp_dir)/$(libsubdir)
     else
--- a/java/Makefile
+++ b/java/Makefile
@@ -20,9 +20,6 @@
 endif
 
 all:
-	$(GRADLE) build
-
-dist:
 	$(GRADLE) :Ice:assemble :Freeze:assemble :Glacier2:assemble :IceGrid:assemble :ant:assemble \
 		:IceBox:assemble :IceDiscovery:assemble :IcePatch2:assemble :IceStorm:assemble :IceGridGUI:assemble
 
--- a/Makefile
+++ b/Makefile
@@ -7,10 +7,10 @@
 #
 # **********************************************************************
 
-SUBDIRS			= cpp java js python ruby php
-CLEAN_SUBDIRS	= js java python ruby php cpp
-DEPEND_SUBDIRS	= cpp python ruby php
-INSTALL_SUBDIRS	= cpp java python ruby php js
+SUBDIRS			= cpp java js php
+CLEAN_SUBDIRS	= js java php cpp
+DEPEND_SUBDIRS	= cpp php
+INSTALL_SUBDIRS	= cpp java php js
 
 ifeq ($(shell uname),Darwin)
 SUBDIRS			+= objective-c
--- a/js/Makefile
+++ b/js/Makefile
@@ -19,6 +19,7 @@
 
 clean: npminstall
 	$(NPM) run gulp:clean
+	rm -rf node_modules
 
 install: npminstall
 	$(NPM) run gulp:install
